src/ValidationExemption.php

```diff
@@ -0,0 +1,247 @@
+	 */
+	public static function mark_node_as_amp_unvalidated( DOMNode $node ) {
+
+		if ( $node instanceof Element ) {
+			if ( ! $node->hasAttribute( self::AMP_UNVALIDATED_TAG_ATTRIBUTE ) ) {
+				if ( ! $node->ownerDocument ) {
+					return false; // @codeCoverageIgnore
+				}
+				$node->setAttributeNode( $node->ownerDocument->createAttribute( self::AMP_UNVALIDATED_TAG_ATTRIBUTE ) );
+			}
+			return true;
+		}
+
+		if ( $node instanceof DOMAttr ) {
+			$element = $node->parentNode;
+			if ( ! $element instanceof Element ) {
+				return false; // @codeCoverageIgnore
+			}
+
+			$attr_value = $element->getAttribute( self::AMP_UNVALIDATED_ATTRS_ATTRIBUTE );
+			if ( $attr_value ) {
+				$attr_value .= ' ' . $node->nodeName;
+			} else {
+				$attr_value = $node->nodeName;
+			}
+
+			$element->setAttribute( self::AMP_UNVALIDATED_ATTRS_ATTRIBUTE, $attr_value );
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Check whether AMP is unvalidated for node.
+	 *
+	 * This means that it is exempted from AMP validation and it may negatively impact PX.
+	 *
+	 * @param DOMNode $node Node.
+	 * @return bool Whether node marked as being AMP-unvalidated.
+	 */
+	public static function is_amp_unvalidated_for_node( DOMNode $node ) {
+
+		// Prevent removing a tag which was exempted from validation.
+		if (
+			$node instanceof \DOMElement
+			&&
+			$node->hasAttribute( self::AMP_UNVALIDATED_TAG_ATTRIBUTE )
+		) {
+			return true;
+		}
+
+		// Check attribute.
+		if (
+			$node instanceof DOMAttr
+			&&
+			self::check_for_attribute_token_list_membership( $node, self::AMP_UNVALIDATED_ATTRS_ATTRIBUTE )
+		) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Check whether the given attribute node is mentioned among the token list of another supplied attribute.
+	 *
+	 * @param DOMAttr $attr                 Attribute node to check for.
+	 * @param string  $token_list_attr_name Attribute name that has the token list value.
+	 *
+	 * @return bool Whether membership is present.
+	 */
+	private static function check_for_attribute_token_list_membership( DOMAttr $attr, $token_list_attr_name ) {
+		$element = $attr->parentNode;
+		if ( ! $element instanceof Element ) {
+			return false;
+		}
+		return (
+			$element->hasAttribute( $token_list_attr_name )
+			&&
+			in_array(
+				$attr->nodeName,
+				preg_split( '/\s+/', $element->getAttribute( $token_list_attr_name ) ),
+				true
+			)
+		);
+	}
+
+	/**
+	 * Check whether the provided document has nodes which are not AMP-validated.
+	 *
+	 * @param Document $document Document for which to check whether dev mode is active.
+	 * @return bool Whether the document is in dev mode.
+	 */
+	public static function is_document_with_amp_unvalidated_nodes( Document $document ) {
+		foreach ( [ self::AMP_UNVALIDATED_TAG_ATTRIBUTE, self::AMP_UNVALIDATED_ATTRS_ATTRIBUTE ] as $attr_name ) {
+			if ( $document->xpath->query( "//*/@{$attr_name}" )->length > 0 ) {
+				return true;
+			}
+		}
+		return false;
```

Comment:
This probably iterates over the entire document twice. Performance-wise, this should be tested against a single run that accepts both attributes instead.
(XPath is untested)
```suggestion
		$xpath_query = sprintf(
			'//*/[@%s or %s]',
			self::AMP_UNVALIDATED_TAG_ATTRIBUTE,
			self::AMP_UNVALIDATED_ATTRS_ATTRIBUTE
		);
		return $document->xpath->query( $xpath_query )->length > 0;
```

Reply:
Done in bd54a6f05ae441e3e9bd6eabec7d442671015218